crates/starknet-types-core/src/felt/mod.rs

```diff
@@ -1661,236 +1685,46 @@ mod test {
         assert_de_tokens(
             &Felt::ZERO.compact(),
             &[
-                Token::Seq { len: Some(32) },
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
-                Token::U8(0),
+                Token::Seq { len: Some(2) },
+                Token::U8(1 | 0x80),
```

Comment:
Do we really need this first element containing len and the compressed flag?
can't we just encode `Felt::ONE` as
```
&[
Token::Seq { len: Some(1) },
Token::U8(1),
]
```
And know that because we are using big endian repr, missing bytes are `0`s and just be happy with it?

Reply:
Serialization can be:

```rust
let be_bytes = self.to_bytes_be();
let mut i = match be_bytes.iter().find(|b| b != 0) {
 Some(i) => i,
 None => { // handle Felt::ZERO and return }
}
let mut seq = serializer.serialize_seq(Some(32 - i))?;
while i < 32 {
    seq.serialize_element(be_bytes[i])?;
    i += 1;
}
seq.end();
```

And deserialization:
```rust
let mut buffer = [0; 32];
let mut i = 32;
while i > 0 {
   if let Some(b) = seq.next_element()? {
         buffer[i-1] = b;
   } else {
           break;
   }
  i -= 1;
}
if i == 0 && seq.next_element()?.is_some() {
  // Return error invalid len
}
```

Reply:
Why wouldn't this way simpler impl work?

Reply:
Using `seq.next_element()?` to infer lengths and handle missing bytes as zeros isn’t reliable for all serializers. Some serializers may not support this implicit handling, leading to incorrect deserialization. Explicitly encoding the length and compressed flag ensures that the deserializer knows exactly how many bytes to expect and process, making the format self-descriptive and robust across different implementations.

Reply:
But isn't what we are already doing with `Some(32 - i)`?

And how exactly could it fail?

Reply:
https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_seq

```
The argument is the number of elements in the sequence, which may or may not be computable before the sequence is iterated. Some serializers only support sequences whose length is known up front.
```

We are fulfilling all conditions described there

And nothing is said about what you are describing there: https://docs.rs/serde/latest/serde/de/trait.SeqAccess.html

Can you give examples or documentation to support your point?

Reply:
It's a misunderstanding on my part.
By following your logic we can directly use se/deserialization using slice

### Serialization
```rust
let bytes = self.to_bytes_be();
let first_significant_byte = bytes.iter().position(|&b| b != 0).unwrap_or(32);
serializer.serialize_bytes(&bytes[first_significant_byte..])
```

### Deserialization
```rust
fn visit_bytes<E>(self, bytes: &[u8]) -> Result<Self::Value, E>
where
    E: de::Error,
{
    if bytes.len() > 32 {
        return Err(E::invalid_length(
            bytes.len(),
            &"a byte array of length up to 32",
        ));
    }
    Ok(Felt::from_bytes_be_slice(&bytes))
}
```

Reply:
The idea here we had with @jbcaron was to make a binary format that never exceeds 32 bytes in the worse case (abitrary hash), and compresses it down to a few bytes in the best cases (small felt)

To do that, we think of the input as a byte stream and we use the felt unused bits in the first byte to encode a flag telling us if the felt is compressed or not

When the felt is compressed we treat that first byte as the length, so we read `len` more bytes from the wire -
when it's not compressed, we treat it as the first byte of the hash and complete it by reading 31 mores bytes.

However, we realized there is an issue: using serde `serialize_seq` (or `serialize_bytes`), bincode will prepend the length of the sequence before the felt, and we end up with >32 bytes felts :(
This is not bincode's problem though it's fundamental to how serde works

Taking inspiration from fixed-sized arrays implementation in serde, they are serialized using `serialize_tuple`, which requires the length to be known beforehand... and we don't know the length before the first byte.

That gave me an idea though - which I tried here https://gist.github.com/cchudant/9bbb025c60a9fe23ca56b895fdcff13c
It works by treating the Felt as a tuple of (u8, \<rest\>) with \<rest\> being another tuple that is `DeserializeSeed`ed with the length
Nesting tuples like this has no effect on the actual bytes on the wire at the end, so we are free to do that
The resulting encoding never exceeds 32 bytes as intended and compressed stuff correctly 

@tdelabro what do you think of that?

Reply:
Still, why do you need the flag for?
Get rid of it and if you have 32 elements, it's not compressed. If you have less it is compressed.
The flag doesn't add anything there

Reply:
You need the byte that contain flag+len only when you serialize [Felt].
Then, because the size in bytes of a Felt can change you have to do:
[flag+len, byte, .., byte, flag+len, byte, flag+len, byte, .., byte, ...]

Reply:
So my recomandation would be:
1. impl Serde on Felt the way I described
2. Create public method `serialize_sequence_of_felt_packed()` and `deserialize_sequence_of_felt_packed()` that can be called inside `#[serde(de/serialize_with = "path")]` and implement your self descriptive packed representation here